### PR TITLE
dada2

### PR DESCRIPTION
**Pull request by bernt-matthias**
moved over from here https://github.com/bernt-matthias/mb-galaxy-tools/pulls

TODOs:
- [x] generate yml based workflow with tests
  - workflow moved to IWC

FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

**Comment by bernt-matthias** [comment]
ping @shiltemann @yvanlebras

**Comment by bernt-matthias** [comment]
Seems that the datatypes have are not yet in the Galaxy version used by planemo. They are alreadey in the Galaxy sources: https://github.com/galaxyproject/galaxy/pull/7895 but I don't know if I was in time for 19.05. 

I could backport them if it would help.

The data manager tests should actually work. Are they invoked correctly?

File: tools/dada2/macros.xml [hunk 1]
```diff
@@ -18,7 +18,8 @@ echo $(R --version | grep version | grep -v GNU)", dada2 version" $(R --vanilla
 
     <xml name="stdio">
         <stdio>
-            <regex match="Error: cannot allocate" source="stderr" level="fatal_oom" description="Out of memory error occurred" />
+				<regex match="Error: cannot allocate" source="stderr" level="fatal_oom" description="Out of memory error occurred" />
```
Comment (bgruening) [inline tools/dada2/macros.xml]: @bernt-matthias something is wrong with your editor, can you please use 4 spaces indentation?

Comment (bernt-matthias) [reply tools/dada2/macros.xml]: True. Maybe I find the time to configure my editor properly one day :)

**Comment by loraine-gueguen** [comment]
hi @bernt-matthias  (cc @lecorguille)
Thanks a lot for you great job on dada2! We are very interested to add these tools in our galaxy instance.
Can I help in any way to solve the issues and complete the development ?

**Comment by bernt-matthias** [comment]
Hi @loraine-gueguen. At the moment you can install the tools from the test toolshed (I just made sure that they are at the latest version). I guess a review and eventually approval of this PR would be the best. 

The errors only affect the data manager (aka dadamanager) .. and I did not understand them .. just tried to rebase to the latest master .. maybe this helps. My local tests are successful. Otherwise I will ask the IUC for help.

**Comment by loraine-gueguen** [comment]
Thanks a lot for the testtoolshed update !
I am completely ignorant about dada2, but I can have a look at the "tool wrapper syntax" stuff of the PR if you want (except data manager unfortunately, I know nothing about data manager :-|)

**Comment by bernt-matthias** [comment]
This would already help a lot.

**Comment by bernt-matthias** [comment]
Surprise: v1.12 (not yet in the TTS)

would be really happy to bring this to an end soon. reviews would be great.

**Comment by loraine-gueguen** [comment]
Sorry I didn't have time to review last 2 weeks (urgent projects and days off)... But I'll do it as soon as possible

**Comment by gregvonkuster** [comment]
@bernt-matthias @bgruening  Sorry I'm getting into this late, I have a new customer that needs this tool in Galaxy.  Is there anything I can do to help move this PR along?

**Comment by yvanlebras** [comment]
+1 I can try to test last version "as a biologist"

**Comment by gregvonkuster** [comment]
@bernt-matthias This code is beautiful!  Just a couple of nit-picky comments, but I don't feel that these necessarily warrant changes unless you agree.

1. Could these lines in the 3 tools become a macro? https://github.com/galaxyproject/tools-iuc/pull/2483/files#diff-246b665596cf9b58d86b16d961ebf249R18-R19

2. Maybe remove commented code in some places just to make the code as clean as possible.
- https://github.com/galaxyproject/tools-iuc/pull/2483/files#diff-fd01f02d9f36105c5ecdac04c7fcdf5cR46
- https://github.com/galaxyproject/tools-iuc/pull/2483/files#diff-fd01f02d9f36105c5ecdac04c7fcdf5cR63-R70
- https://github.com/galaxyproject/tools-iuc/pull/2483/files#diff-11050603102d7277e183b7a42d445c61R11
- https://github.com/galaxyproject/tools-iuc/pull/2483/files#diff-a1bc34a4e421210fe3ebb92146b232acR29

I've restricted my review to the tool configs, including data managers, and R code.  If nothing is missing form this PR to make the tools all work within Galaxy (I'm late to the game here), then I'm +1 on this when tests pass.

**Comment by bernt-matthias** [comment]
hey @gregvonkuster thanks for the review. I removed the comments (in one case I decided to leave it, but format it properly). 

I decided against a macro for the 2 lines, I had the feeling that it does not help with the readability of the code.

**Comment by bernt-matthias** [comment]
Anyone problems with `profile="19.09"` (which is not even released yet)? Needed because of the new data types. If one would vote for an older Galaxy release I would need to backport the corresponding commit.

**Comment by bernt-matthias** [comment]
The most important question to me is how well the output of the pipeline (i.e. of assignTaxonomy  / makeSequenctable) as implemented in the wrapppers
- can be fed to connect to possible downstream tools
- are consistent with other amplicon approaches in Galaxy (qiime / mothur)

Where the first seems more important to me.

**Comment by gregvonkuster** [comment]
> Anyone problems with `profile="19.09"` (which is not even released yet)? Needed because of the new data types. If one would vote for an older Galaxy release I would need to backport the corresponding commit.

I'll need this on Galaxy main which states that it is currently running 19.09, but not sure since it is not yet released.  ;)

File: tools/dada2/datatypes_conf.xml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
```
Comment (mvdbeek) [inline tools/dada2/datatypes_conf.xml]: Should this still be part of this PR ?

Comment (bernt-matthias) [reply tools/dada2/datatypes_conf.xml]: Nope. Consider it gone.

**Comment by bernt-matthias** [comment]
> The fastq.gz files are relatively large, any chance that you could downsample them a bit more ?

Consider it done. Do I need to rebase and squash the commits to actually affect the size of the git repo?

**Comment by nsoranzo** [comment]
> Do I need to rebase and squash the commits to actually affect the size of the git repo?

That would be better, otherwise whoever merges this PR will need to remember to select "Squash and merge".

**Comment by bgruening** [comment]
Just send my two last comments. Please check the tool owners ... I can squash during merge.

**Comment by bernt-matthias** [comment]
Will be a few days offline and implement the changes afterwards.

I think squashing is not necessary anymore - I rebased already.

File: data_managers/data_manager_dada2/.shed.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+name: data_manager_dada2
+owner: matthias
+description: "Data manager to download DADA2 reference databases"
+homepage_url: "https://benjjneb.github.io"
```
Comment (loraine-gueguen) [inline data_managers/data_manager_dada2/.shed.yml]: `homepage_url: "https://benjjneb.github.io/dada2/index.html"`
The url is broken

File: data_managers/data_manager_dada2/.shed.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+long_description: |
+  DADA2: Fast and accurate sample inference from amplicon data with single-nucleotide resolution
+remote_repository_url: "https://github.com/bernt-matthias/mb-galaxy-tools/tree/master/data_managers/data_manager_dada2"
```
Comment (loraine-gueguen) [inline data_managers/data_manager_dada2/.shed.yml]: Won't it be this repository instead ?
`remote_repository_url: "https://github.com/galaxyproject/tools-iuc/tree/master/data_managers/data_manager_dada2"`

File: tools/dada2/.shed.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+name: dada2
+owner: iuc
+categories:
+  - Metagenomics
+description: DADA2 wrappers
+long_description: |
+    Fast and accurate sample inference from amplicon data with single-nucleotide resolution
+categories:
+- Metagenomics
+remote_repository_url: https://github.com/bernt-matthias/mb-galaxy-tools/tree/topic/dada2/tools/dada2
```
Comment (loraine-gueguen) [inline tools/dada2/.shed.yml]: Won't it be this repository instead ?
`remote_repository_url: https://github.com/galaxyproject/tools-iuc/tree/master/tools/dada2`

File: tools/dada2/dada2_filterAndTrim.xml [hunk 6]
```diff
@@ -0,0 +1,264 @@
+<tool id="dada2_filterAndTrim" name="dada2: filterAndTrim" version="@DADA2_VERSION@+galaxy@WRAPPER_VERSION@" profile="19.09">
+    <description>Filters and trims an short read data</description>
```
Comment (loraine-gueguen) [inline tools/dada2/dada2_filterAndTrim.xml]: typos ?
`    <description>Filter and trim short read data</description>`

**Comment by loraine-gueguen** [comment]
We have a few interested users that will test the suite. Is the code strictly limited to installation and use with release 19.09 ?

**Comment by bernt-matthias** [comment]
>  Is the code strictly limited to installation and use with release 19.09 ?

The only aspect that requires 19.09 are the data types. You can add them manually to your datatypes.xml. I'm not sure if installation from the TS is checking for the version, but manual installation is definitely possible.

Let me know if I should update the version in the TTS.

**Comment by loraine-gueguen** [comment]
It seems not possible to install a tool with a profile higher than the current galaxy profile, neither from the TS or manually (unless editing every tool config to change the profile, what I turned out to do).
 
From the TS, the tools appear as "invalid" in the admin "manage tools" section with no error messages. Manually, I had the following error messages:
```
  File "lib/galaxy/tools/__init__.py", line 446, in __init__
    raise e
Exception: The tool dada2_assignTaxonomyAddspecies targets version 19.09 of Galaxy, you should upgrade Galaxy to ensure proper functioning of this tool.

File: data_managers/data_manager_dada2/.shed.yml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+name: data_manager_dada2
+owner: matthias
```
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply data_managers/data_manager_dada2/.shed.yml]: done

File: data_managers/data_manager_dada2/data_manager/dada2_fetcher.xml [hunk 8]
```diff
@@ -0,0 +1,181 @@
+<?xml version="1.0"?>
+<tool id="dada_fetcher" name="dada2 data manager" tool_type="manage_data" version="0.0.7">
+    <description>Download reference databases</description>
+    <command detect_errors="exit_code"><![CDATA[
+    python '$__tool_directory__/data_manager.py'
+	--out '$out_file'
+	#set dataset = str($db_cond.db_select) + '_' + str($db_cond.version_select)
+    --dataset '$dataset'
```
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply data_managers/data_manager_dada2/data_manager/dada2_fetcher.xml]: done

File: data_managers/data_manager_dada2/.shed.yml (hunk 3 above)
Comment (bernt-matthias) [reply data_managers/data_manager_dada2/.shed.yml]: URL seems to work. Maybe it was a temporary issue.. just noticed that another upgrade of dada2 is out now .. which needs to wait.

File: data_managers/data_manager_dada2/.shed.yml (hunk 4 above)
Comment (bernt-matthias) [reply data_managers/data_manager_dada2/.shed.yml]: done

File: tools/dada2/.shed.yml (hunk 5 above)
Comment (bernt-matthias) [reply tools/dada2/.shed.yml]: done

File: tools/dada2/dada2_filterAndTrim.xml (hunk 6 above)
Comment (bernt-matthias) [reply tools/dada2/dada2_filterAndTrim.xml]: good catch. done

**Comment by bernt-matthias** [comment]
@loraine-gueguen good to know. manually editing the tool xml files should boil down to something like: 

```
find . -name "dada2_*xml" -exec sed -i -e 's/profile="19.09"/profile="19.01"/' {} ;
```

and adding the data types. 

Could be added to the docs.

File: data_managers/data_manager_dada2/.shed.yml (hunk 3 above)
Comment (loraine-gueguen) [reply data_managers/data_manager_dada2/.shed.yml]: strange: still 404 error code for me with https://benjjneb.github.io/

Comment (bernt-matthias) [reply data_managers/data_manager_dada2/.shed.yml]: changed to https://benjjneb.github.io/dada2/index.html

**Comment by bgruening** [comment]
Ok, let's get this in, way too long open. Hope that is ok for everyone ...

**Comment by loraine-gueguen** [comment]
A few comments regarding our DADA2  installation in Galaxy from the TS:

For Galaxy releases older than 19.09, it seems not enough to manually change profile in the xml files (restrictions due to a profile parameter registered in the database ?). To make the tools available in our 18.05 instance, I had to add the tools manually in `galaxy-dist/tools/dada2/` (in fact, I added symbolic links in pointing to the repositories of each tool in `shed_tools`). The best would be to update our galaxy instance to 19.09, which is now released.

Moreover, a release of conda >4.5.11 is definitely needed to have the bioconductor-dada2@1.12 conda environment installing correctly: I faced the same issue as https://github.com/bioconda/bioconda-recipes/issues/13846#issuecomment-485841543, and it was ok after updating conda to 4.712 (and python).

Small stuffs: the image pairpipe.png is not displaying in the help section (`pairpipe.png` instead), and the doi citation displays an error message `Failed to fetch BibTeX for DOI` (although the hyperlinks are ok). Do you want me to open an issue for these ?

A user is currently testing the tool suite.

**Comment by gregvonkuster** [comment]
This tool suite is also currently being tested on Galaxy test https://test.galaxyproject.org, although the reference datasets are not yet installed for the tools that use them.

**Comment by bernt-matthias** [comment]
> For Galaxy releases older than 19.09, it seems not enough to manually change profile in the xml files (restrictions due to a profile parameter registered in the database ?). To make the tools available in our 18.05 instance, I had to add the tools manually in galaxy-dist/tools/dada2/ (in fact, I added symbolic links pointing to the repositories of each tool in shed_tools). The best would be to update our galaxy instance to 19.09, which is now released.

Odd. Do the dada tools end up in shed_tools.xml? Or in which of the tool.xml files?

> Moreover, a release of conda >4.5.11 is definitely needed to have the bioconductor-dada2@1.12 conda environment installing correctly: I faced the same issue as bioconda/bioconda-recipes#13846 (comment), and it was ok after updating conda to 4.7.12 (and python).

A small comment on the README in the repo would be nice. Could you do this?

> Small stuffs: the image pairpipe.png is not displaying in the help section (pairpipe.png instead), and the doi citation displays an error message Failed to fetch BibTeX for DOI (although the hyperlinks are ok). Do you want me to open an issue for these ?

Images seem to work only when installed from the TS.

**Comment by bernt-matthias** [comment]
@gregvonkuster cool :) .. manual tests I guess?

**Comment by gregvonkuster** [comment]
A lab here on the Penn State campus is running some of the tools on their data.  If all goes well, I'll try to get the tools installed on Galaxy main before the Christmas break or shortly after the new year.

**Comment by gregvonkuster** [comment]
I need to get in touch with @jennaj to get the reference datasets installed.

**Comment by gregvonkuster** [comment]
This @bernt-matthias I've opened this issue https://github.com/galaxyproject/usegalaxy-playbook/issues/273 requesting dada2 reference databases to be installed on Galaxy test / main, but I'm not quite sure which databases are the most used.  So if I've chosen incorrectly, please let me know.

**Comment by loraine-gueguen** [comment]
> Odd. Do the dada tools end up in shed_tools.xml? Or in which of the tool.xml files?
The dada tools were correctly added in a shed_tool_conf file dedicated to dada2 tools (`shed_tool_conf_dada2.xml`) which has been added in the `tool_config_file` parameter in `galaxy.yml`. And the installation went fine for dada2 v1.10 from the testtoolshed in the same shed tool_conf file.

> A small comment on the README in the repo would be nice. Could you do this?
done: https://github.com/galaxyproject/tools-iuc/pull/2699

> Images seem to work only when installed from the TS.
ok. For DOI citation as well ?

Thanks

**Comment by bernt-matthias** [comment]
Any immediate requests for dada2 could be added here: https://github.com/galaxyproject/tools-iuc/pull/2705

**Pull request merged**